api/src/main/java/org/geysermc/geyser/api/item/custom/v2/CustomItemDefinition.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2024 GeyserMC. http://geysermc.org
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.item.custom.v2;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.geysermc.geyser.api.GeyserApi;
+import org.geysermc.geyser.api.item.custom.v2.component.DataComponentMap;
+import org.geysermc.geyser.api.item.custom.v2.component.DataComponent;
+import org.geysermc.geyser.api.item.custom.v2.predicate.CustomItemPredicate;
+import org.geysermc.geyser.api.item.custom.v2.predicate.PredicateStrategy;
+import org.geysermc.geyser.api.util.Identifier;
+
+import java.util.List;
+
+/**
+ * This is used to define a custom item and its properties for a specific Java item and item model definition combination.
+ *
+ * <p>A custom item definition will be used for all item stacks that match the Java item and item model this item is for.
+ * Additionally, predicates can be added that allow fine-grained control as to when to use this custom item. These predicates are similar
+ * to the predicates available in Java item model definitions.</p>
+ *
+ * <p>In Geyser, all registered custom item definitions for a Java item model will be checked in a specific order:</p>
+ *
+ * <ol>
+ *     <li>First by checking their priority values, higher priority values going first.</li>
+ *     <li>Then by checking if they both have a similar range dispatch predicate, the one with the highest threshold going first.</li>
+ *     <li>Lastly by the amount of predicates, from most to least.</li>
+ * </ol>
+ *
+ * <p>This ensures predicates will be checked in a correct order, and that in most cases specifying a priority value isn't necessary, and in the few cases
+ * where Geyser doesn't properly sort definitions, specifying a priority value will.</p>
```

Comment:
```suggestion
 * <p>This ensures predicates will be checked in the correct order. In most cases, specifying a priority value isn't necessary, but it can be added to ensure the intended order. </p>
```